sql: support `COMMENT ON DATABASE`

hueypark opened the pull request:
Informs #19472.

This patch introduces support for database comments.

The syntax to set or delete a comment is the same as postgres:
`COMMENT ON DATABASE ... IS ...`. See:
https://www.postgresql.org/docs/9.1/sql-comment.html

Release note (sql change): CockroachDB now supports associating
comments to SQL databases using PostgreSQL's `COMMENT ON DATABASE`
syntax. This also provides proper support for pg's
`pg_catalog.pg_description` and built-in function `obj_description()`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/33057)
<!-- Reviewable:end -->

knz commented:
@hueypark thank you for your PR! I am _so_ sorry we did not notice it earlier. I am so excited that you have prepared this already!

We will review it shortly. Thank you again.

Meanwhile my colleague @rolandcrosby is curious to know whether you also would like to work on column comments? I am not sure what to say (of course it's great if you would like to, but we can also do this work ourselves afterwards).

hueypark commented:
@knz, @rolandcrosby 
Of course, I'll work on the column next.

knz reviewed (commented):
Hi again, sorry for the late review. Here are my comments.

1) the code is overall very good like last time, very well done

2) I just noticed a problem here which also exists in your previous code for COMMENT ON TABLE: you have added event names for the event logger (EventLogCommentOnDatabase etc) but you did not actually create events.

I think the following code is missing in a couple of places in the `commonOn...Node` code:

`MakeEventLogger(params.extendedEvalCtx.ExecCfg).InsertEventRecord(...EventLogCommentOn....)`

Thank you

> Reviewed 17 of 18 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33057#-:-LUp6mMc7OiEHt3GVK9W:b1oh44f)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/comment_on_database.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33057#-LUp5usl78QKbG3e6MD_:-LUp5usl78QKbG3e6MDa:bb7mfsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f74e723a21ba427b496c376b48830b62749cdcad/pkg/sql/comment_on_database.go#L32)):*
> ```Go
> 
> // CommentOnDatabase add comment on a database.
> // Privileges: CREATE on table.
> ```

this 2nd line of the comment is incorrect -- the privilege check is also on the database here.
___
*[pkg/sql/comment_on_database.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33057#-LUp65ElDEFuwStFykNP:-LUp65ElDEFuwStFykNQ:bf96vt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f74e723a21ba427b496c376b48830b62749cdcad/pkg/sql/comment_on_database.go#L53)):*
> ```Go
> 		_, err := params.p.extendedEvalCtx.ExecCfg.InternalExecutor.Exec(
> 			params.ctx,
> 			"upsert-comment",
> ```

you can use "set-db-comment" here. You can also change the other one for COMMENT ON TABLE to become "set-table-comment"

The word "upsert" is an implementation detail. We also want to distinguish the events in traces.
___
*[pkg/sql/comment_on_database.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33057#-LUp6HsZ5ENtKvdvc5p1:-LUp6HsZ5ENtKvdvc5p2:b-dx1d72) ([raw file](https://github.com/cockroachdb/cockroach/blob/f74e723a21ba427b496c376b48830b62749cdcad/pkg/sql/comment_on_database.go#L65)):*
> ```Go
> 		_, err := params.p.extendedEvalCtx.ExecCfg.InternalExecutor.Exec(
> 			params.ctx,
> 			"delete-comment",
> ```

ditto `delete-db-comment` and change the table one to `delete-table-comment`


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 1 of 18 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33057)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

hueypark committed:
sql: support `COMMENT ON DATABASE`

This patch introduces support for database comments.

The syntax to set or delete a comment is the same as postgres:
`COMMENT ON DATABASE ... IS ...`. See:
https://www.postgresql.org/docs/9.1/sql-comment.html

Release note (sql change): CockroachDB now supports associating
comments to SQL databases using PostgreSQL's `COMMENT ON DATABASE`
syntax. This also provides proper support for pg's
`pg_catalog.pg_description` and built-in function `obj_description()`.

hueypark reviewed (commented):
I am already calling `MakeEventLogger`. Please check again.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33057#-:-LUs_Srj7Q0j2kU7Sh1y:b8xv6te)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/comment_on_database.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33057#-LUp5usl78QKbG3e6MD_:-LUs__GcBtyqyEqzQjFo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f74e723a21ba427b496c376b48830b62749cdcad/pkg/sql/comment_on_database.go#L32)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

this 2nd line of the comment is incorrect -- the privilege check is also on the database here.
</blockquote></details>

Done.
___
*[pkg/sql/comment_on_database.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33057#-LUp65ElDEFuwStFykNP:-LUs_ayMDn4Bese9PJDb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f74e723a21ba427b496c376b48830b62749cdcad/pkg/sql/comment_on_database.go#L53)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

you can use "set-db-comment" here. You can also change the other one for COMMENT ON TABLE to become "set-table-comment"

The word "upsert" is an implementation detail. We also want to distinguish the events in traces.
</blockquote></details>

Done.
___
*[pkg/sql/comment_on_database.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33057#-LUp6HsZ5ENtKvdvc5p1:-LUs_cXc7lIWqC7J-Ac_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f74e723a21ba427b496c376b48830b62749cdcad/pkg/sql/comment_on_database.go#L65)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto `delete-db-comment` and change the table one to `delete-table-comment`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> I am already calling MakeEventLogger. Please check again.

Yes I see it now. Thank you.

> Reviewed 11 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33057#-:-LUzLnhj6ycEDbVObrvr:b-3knlb7)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

knz commented:
Very nice. Thank you for your contribution!

bors r+
